# Opening the file
---------------------
Here we turn a large dataset into a list of list

In [108]:
opened_file = open('StudentsPerformance.csv')
from csv import reader
read_file = reader(opened_file)
student_data =list(read_file)


# Counting with Dictionaries


### Gender Proportions
First we determine gender proportions by answering the following questions:

a.What is the proportion of male and female students in the dataset?

b. Is there any gender imbalance among students' performance in math, reading, and writing? If so, what are the proportions?


In [109]:
Gender = {}
for row in student_data[1:]:
    unique_gender = row[0]
    if unique_gender in Gender:
        Gender[unique_gender] += 1
    else:
        Gender[unique_gender] = 1
        
print(Gender)

{'female': 518, 'male': 482}


In [110]:
total_student = len(student_data)


male_proportion = Gender['male'] / total_student
female_proportion = Gender['female'] / total_student

male_percentage = male_proportion * 100
female_percentage = female_proportion * 100

print(f"Proportion of male students: {male_proportion:.2f}")
print(f"Proportion of female students: {female_proportion:.2f}")

Proportion of male students: 0.48
Proportion of female students: 0.52


### Hence,the proportion of male student in the dataset is given to be 48% while that of the female is 52%

In [111]:
male_math_scores = []
female_math_scores = []
for row in student_data[1:]:
    gender = row[0]
    math_score = int(row[5])
    if gender == 'male':
        male_math_scores.append(math_score)
    else:
        female_math_scores.append(math_score)
        
math_score_tot = ((sum(male_math_scores)) + sum(female_math_scores))

print(sum(male_math_scores))
print(sum(female_math_scores))


33127
32962


In [112]:
male_reading_scores = []
female_reading_scores = []
for row in student_data[1:]:
    gender = row[0]
    reading_score = int(row[6])
    if gender == 'male':
        male_reading_scores.append(reading_score)
    else:
        female_reading_scores.append(reading_score)

reading_score_tot = ((sum(male_reading_scores)) + sum(female_reading_scores))
print(sum(male_reading_scores))
print(sum(female_reading_scores))

31558
37611


In [113]:
male_writing_scores = []
female_writing_scores = []
for row in student_data[1:]:
    gender = row[0]
    writing_score = int(row[7])
    if gender == 'male':
        male_writing_scores.append(writing_score)
    else:
        female_writing_scores.append(writing_score)
        
writing_score_tot = ((sum(male_writing_scores)) + sum(female_writing_scores))

print(sum(male_writing_scores))
print(sum(female_writing_scores))

30516
37538


In [114]:
math_imbalance = ((sum(female_math_scores) - sum(male_math_scores))  / math_score_tot) 
reading_imbalance = ((sum(female_reading_scores) -sum(male_reading_scores))  / reading_score_tot) 
writing_imbalance = ((sum(female_writing_scores) -sum(male_writing_scores))  / writing_score_tot) 

print(f"The gender imbalance in Maths is: {math_imbalance:.2f}")
print(f"The gender imbalance in Reading is: {reading_imbalance:.2f}")
print(f"The gender imbalance in Writing is: {writing_imbalance:.2f}")


The gender imbalance in Maths is: -0.00
The gender imbalance in Reading is: 0.09
The gender imbalance in Writing is: 0.10


### Therefore the gender imbalance  among students' performance in Maths,Reading and Writing based on their proportions are -0.00, 0.09, 0.10

------------
### Race/Ethnicity Proportions

a.What are the proportions of different race/ethnicity groups in the dataset?

b. Determine the percentage of different race/ethnicity groups in the dataset ?


In [115]:
ethic_groups = {}
for row in student_data[1:]:
    race =row[1]
    if race in ethic_groups:
        ethic_groups[race] += 1
    else:
        ethic_groups[race] = 1
             
ethic_groups


{'group B': 190, 'group C': 319, 'group A': 89, 'group D': 262, 'group E': 140}

In [116]:
ethic_groups_proportion = {}
ethic_groups_percentage ={}
for row in ethic_groups:
    proportion = ethic_groups[row] / total_student
    percentage = proportion * 100
    
    ethic_groups_proportion[row] = proportion
    ethic_groups_percentage[row] = percentage
    
    
ethic_prop_dict = {k: round(v, 2) for k, v in   ethic_groups_proportion .items()}
ethic_percent_dict = {k: round(v, 2) for k, v in   ethic_groups_percentage .items()}

print(ethic_prop_dict)
print(ethic_percent_dict)

{'group B': 0.19, 'group C': 0.32, 'group A': 0.09, 'group D': 0.26, 'group E': 0.14}
{'group B': 18.98, 'group C': 31.87, 'group A': 8.89, 'group D': 26.17, 'group E': 13.99}


### The proportions of different race/ethicity in the dataset is given as :
Group A : 0.09, 
Group B : 0.19, 
Group C : 0.32, 
Group D : 0.26, 
Group E : 0.14, 

---------
### The percentages of different race/ethicity in the dataset is given as :
Group A : 8.89, 
Group B : 18.98, 
Group C : 31.87, 
Group D : 26.17, 
Group E : 13.99, 

------------


### Parental Level of Education:

a. What are the proportions of students with different levels of parental education?


In [117]:
parental_level = {}
for row in student_data[1:]:
    level =row[2]
    if level in parental_level:
        parental_level[level] += 1
    else:
        parental_level[level] = 1
             
parental_level

{"bachelor's degree": 118,
 'some college': 226,
 "master's degree": 59,
 "associate's degree": 222,
 'high school': 196,
 'some high school': 179}

In [118]:
parental_level_proportion = {}
parental_level_percentage = {}

for row in parental_level:
    proportion = parental_level[row] / total_student
    percentage = proportion * 100
    
    parental_level_proportion[row] = proportion
    parental_level_percentage[row] = percentage
    
    
parental_level_prop_dict = {k: round(v, 2) for k, v in   parental_level_proportion .items()}
parental_level_percent_dict = {k: round(v, 2) for k, v in   parental_level_percentage .items()}

print(parental_level_prop_dict)
print(parental_level_percent_dict)

{"bachelor's degree": 0.12, 'some college': 0.23, "master's degree": 0.06, "associate's degree": 0.22, 'high school': 0.2, 'some high school': 0.18}
{"bachelor's degree": 11.79, 'some college': 22.58, "master's degree": 5.89, "associate's degree": 22.18, 'high school': 19.58, 'some high school': 17.88}


### The proportions of different race/ethicity in the dataset is given as :

 "Bachelor's degree": 0.12, 'Some college': 0.23, "Master's degree": 0.06, "Associate's degree": 0.22, 'High school': 0.2, 
 'Some high school': 0.18
 
 ----------

-------
-------
# Creating Frequency Tables With Intervals

## Evaluating student's performance in Maths

First we determine the minimum and maximum scores to enable us create reasonable intervals

In [119]:
maths_scores = []
for row in student_data[1:]:
    score = int(row[5] )
    maths_scores.append(score)

min_score = min(maths_scores)
max_score = max(maths_scores)

print(min_score)
print(max_score)


0
100


------
Here we evaluate the student's performance in maths by allocating maths scores to a five point interval to see their performances ranging from "Very Poor" to "Excellent"

In [120]:
maths_grades = {"Very Poor" : 0, "Poor":0, "Average" : 0, "Good": 0, "Excellent" : 0}

for row in student_data[1:]:
    score = int(row[5])
    
    if score < 19:
        maths_grades["Very Poor"] += 1
    elif 20 < score <= 39:
        maths_grades["Poor"] += 1
    elif 50 < score <= 59:
        maths_grades["Average"] += 1
    elif 60 < score <= 79: 
        maths_grades["Good"] += 1
    elif score > 80:
        maths_grades["Excellent"] += 1
        
maths_grades

{'Very Poor': 3, 'Poor': 36, 'Average': 173, 'Good': 468, 'Excellent': 176}

In [121]:
maths_grade_proportion = {}
maths_grade_percentage = {}

for row in maths_grades:
    proportion = maths_grades[row] / total_student
    percentage = proportion * 100
    
    maths_grade_proportion[row] = proportion
    maths_grade_percentage[row] = percentage
    
    
maths_grade_prop_dict = {k: round(v, 2) for k, v in   maths_grade_proportion .items()}
maths_grade_percent_dict = {k: round(v, 0) for k, v in   maths_grade_percentage .items()}

print(maths_grade_prop_dict)
print(maths_grade_percent_dict)

{'Very Poor': 0.0, 'Poor': 0.04, 'Average': 0.17, 'Good': 0.47, 'Excellent': 0.18}
{'Very Poor': 0.0, 'Poor': 4.0, 'Average': 17.0, 'Good': 47.0, 'Excellent': 18.0}


### Therefore , the evaluation of the student's performance in maths in the dataset show that:

0% of the students had "Very poor" grades

4% of the students had "Poor" grades

17% of the students  had "Average" grades

47% of the students had "Good" grades

18% of the students had "Excellent" grades


---
---

## Evaluating student's Reading Habits

First we determine the minimum and maximum values to enable us create reasonable intervals

In [122]:
reading_scores = []
for row in student_data[1:]:
    score = int(row[6] )
    reading_scores.append(score)

min_score = min(reading_scores)
max_score = max(reading_scores)

print(min_score)
print(max_score)



17
100


#### Here student's reading habits are measured on a score of 0 - 100 with a minimum value of 17 and maximum value of 100

In [123]:
reading_habit = {"Minimal Effort" : 0, "Below Average":0, "Average" : 0, "Above Average": 0, "Exceptional Effort" : 0}

for row in student_data[1:]:
    score = int(row[6])
    
    if score < 19:
        reading_habit["Minimal Effort"] += 1    
    elif 20 < score <= 39:
        reading_habit["Below Average"] += 1
    elif 50 < score <= 59:
        reading_habit["Average"] += 1
    elif 60 < score <= 79: 
        reading_habit["Above Average"] += 1
    elif score > 80:
        reading_habit["Exceptional Effort"] += 1
        
reading_habit

{'Minimal Effort': 1,
 'Below Average': 25,
 'Average': 157,
 'Above Average': 476,
 'Exceptional Effort': 235}

In [124]:
reading_habit_proportion = {}
reading_habit_percentage = {}

for row in reading_habit:
    proportion = reading_habit[row] / total_student
    percentage = proportion * 100
    
    reading_habit_proportion[row] = proportion
    reading_habit_percentage[row] = percentage
    
    
reading_habit_prop_dict = {k: round(v, 2) for k, v in   reading_habit_proportion .items()}
reading_habit_percent_dict = {k: round(v, 0) for k, v in   reading_habit_percentage.items()}

print(reading_habit_prop_dict)
print(reading_habit_percent_dict )

{'Minimal Effort': 0.0, 'Below Average': 0.02, 'Average': 0.16, 'Above Average': 0.48, 'Exceptional Effort': 0.23}
{'Minimal Effort': 0.0, 'Below Average': 2.0, 'Average': 16.0, 'Above Average': 48.0, 'Exceptional Effort': 23.0}


### Therefore , the evaluation of the student's performance in maths in the dataset show that:
0% of the students put "Minimal Effort" in studying

2% of the students had "Below Average" studying habit

16% of the students had "Average" studying habit

48% of the students had "Above Average" studying habit

18% of the students put "Exceptional Effort" in studying
